File: apps/mcp-server/src/infrastructure/fetchers/newsapi.ts
/**
 * Infrastructure — NewsAPI.org Fetcher
 *
 * Fetches Vietnam-related news from NewsAPI.org as a fallback source when
 * the VPS Reuters push has been absent for >90 minutes.
 *
 * Safe no-key path: returns [] immediately when apiKey is empty or enabled
 * is false — zero I/O, suitable for deployments without a NewsAPI key.
 *
 * API docs: https://newsapi.org/docs/endpoints/everything
 *
 * Layer: infrastructure/fetchers — may use HTTP libs, must not import domain/.
 *
 * @module infrastructure/fetchers/newsapi
 */

import type { RssItem } from "./rss.js";
import { logger } from "../logger.js";

// ─────────────────────────────────────────────────────────────────────────────
// Constants
// ─────────────────────────────────────────────────────────────────────────────

/** NewsAPI endpoint for "everything" search. */
const NEWSAPI_URL = "https://newsapi.org/v2/everything";

/** Default query: Vietnam economy and stock market news. */
const DEFAULT_QUERY = "Vietnam economy OR Vietnam stock market OR VN-Index";

/** Max results per request — NewsAPI free tier allows up to 100. */
const PAGE_SIZE = 20;

/** Source tag applied to all items from this fetcher. */
const SOURCE_TAG = "newsapi";

// ─────────────────────────────────────────────────────────────────────────────
// Types
// ─────────────────────────────────────────────────────────────────────────────

/** Configuration injected from mcpConfig.newsSources.newsapi. */
export interface NewsApiConfig {
  /** NewsAPI.org API key. Empty string = stub path (returns []). */
  apiKey: string;
  /** Master switch — false means skip entirely. */
  enabled: boolean;
  /** Optional query override (defaults to DEFAULT_QUERY). */
  query?: string;
}

/** Shape of one article from the NewsAPI /everything response. */
interface NewsApiArticle {
  title: string | null;
  url: string | null;
  publishedAt: string | null;
  description: string | null;
  source?: { name?: string };
}

// ─────────────────────────────────────────────────────────────────────────────
// Public API
// ─────────────────────────────────────────────────────────────────────────────

/**
 * Fetches Vietnam-related news from NewsAPI.org.
 *
 * Stub path: returns [] immediately when `apiKey` is empty or `enabled` is
 * false — zero network calls, safe for deployments without a key.
 *
 * @param config  - NewsAPI configuration (key, enabled flag, optional query)
 * @returns       - Array of RssItem (empty on failure or stub path)
 */
export async function fetchNewsApi(config: NewsApiConfig): Promise<RssItem[]> {
  // Stub path — no key or disabled: zero I/O
  if (!config.enabled || !config.apiKey) {
    logger.debug("[newsapi] skipping fetch — disabled or no apiKey configured");
    return [];
  }

  const query = config.query ?? DEFAULT_QUERY;
  const url = new URL(NEWSAPI_URL);
  url.searchParams.set("q", query);
  url.searchParams.set("language", "en");
  url.searchParams.set("sortBy", "publishedAt");
  url.searchParams.set("pageSize", String(PAGE_SIZE));
  url.searchParams.set("apiKey", config.apiKey);

  try {
    logger.debug("[newsapi] fetching articles", { query, pageSize: PAGE_SIZE });

    const response = await fetch(url.toString(), {
      headers: {
        "User-Agent":
          "Mozilla/5.0 (compatible; VNMarket/1.0; +https://github.com/phamhung075/VN-Market-Intelligence-MCP)",
        Accept: "application/json",
      },
      signal: AbortSignal.timeout(15_000),
    });

    if (!response.ok) {
      logger.warn("[newsapi] HTTP error", { status: response.status });
      return [];
    }

    const data = (await response.json()) as {
      status: string;
      articles?: NewsApiArticle[];
    };

    if (data.status !== "ok" || !Array.isArray(data.articles)) {
      logger.warn("[newsapi] unexpected response shape", { status: data.status });
      return [];
    }

    const items: RssItem[] = data.articles
      .filter((a) => a.title && a.url)
      .map((a) => ({
        title: a.title!,
        url: a.url!,
        source: SOURCE_TAG,
        publishedAt: a.publishedAt ?? new Date().toISOString(),
        content: a.description ?? "",
      }));

    logger.info("[newsapi] fetched articles", { count: items.length });
    return items;
  } catch (err) {
    logger.error("[newsapi] fetch failed", {
      error: err instanceof Error ? err.message : String(err),
    });
    return [];
  }
}

File: apps/news-fetch/internal/fetcher/newsapi.go
// Package fetcher — NewsAPI.org integration for Vietnam-related news.
//
// Stub path: returns nil immediately when apiKey is empty or enabled=false.
// Zero I/O on the stub path — safe for deployments without a NewsAPI key.
//
// API docs: https://newsapi.org/docs/endpoints/everything
package fetcher

import (
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"time"
)

// NewsAPIConfig holds the configuration for the NewsAPI fetcher.
type NewsAPIConfig struct {
	// APIKey is the newsapi.org API key. Empty string = stub path (returns nil).
	APIKey string
	// Enabled is a master switch. false = skip entirely, zero I/O.
	Enabled bool
	// Query overrides the default Vietnam news query when non-empty.
	Query string
}

// newsAPIResponse mirrors the top-level JSON from newsapi.org /v2/everything.
type newsAPIResponse struct {
	Status   string           `json:"status"`
	Articles []newsAPIArticle `json:"articles"`
}

// newsAPIArticle mirrors one article from the NewsAPI /everything response.
type newsAPIArticle struct {
	Title       string      `json:"title"`
	URL         string      `json:"url"`
	PublishedAt string      `json:"publishedAt"`
	Description string      `json:"description"`
	Source      *newsSource `json:"source"`
}

type newsSource struct {
	Name string `json:"name"`
}

const (
	newsAPIURL      = "https://newsapi.org/v2/everything"
	defaultQuery    = "Vietnam economy OR Vietnam stock market OR VN-Index"
	newsAPIPageSize = 20
	newsAPISource   = "newsapi"
)

// NewsAPIFetcher fetches Vietnam-related news from newsapi.org.
type NewsAPIFetcher struct {
	cfg    NewsAPIConfig
	client HTTPClient
}

// NewNewsAPIFetcher creates a NewsAPIFetcher with the given config.
// client may be nil to use the default production HTTP client.
func NewNewsAPIFetcher(cfg NewsAPIConfig, client HTTPClient) *NewsAPIFetcher {
	if client == nil {
		client = newDefaultHTTPClient(15 * time.Second)
	}
	return &NewsAPIFetcher{cfg: cfg, client: client}
}

// Fetch fetches Vietnam-related articles from newsapi.org.
// Returns nil, nil on the stub path (disabled or no API key).
func (f *NewsAPIFetcher) Fetch(ctx context.Context) ([]RssItem, error) {
	if !f.cfg.Enabled || f.cfg.APIKey == "" {
		// Stub path — zero I/O
		return nil, nil
	}

	query := f.cfg.Query
	if query == "" {
		query = defaultQuery
	}

	params := url.Values{}
	params.Set("q", query)
	params.Set("pageSize", fmt.Sprintf("%d", newsAPIPageSize))
	params.Set("language", "en")
	params.Set("sortBy", "publishedAt")
	params.Set("apiKey", f.cfg.APIKey)

	reqURL := newsAPIURL + "?" + params.Encode()

	req, err := http.NewRequestWithContext(ctx, http.MethodGet, reqURL, nil)
	if err != nil {
		return nil, fmt.Errorf("newsapi: build request: %w", err)
	}
	req.Header.Set("User-Agent", browserUA)

	resp, err := f.client.Do(req)
	if err != nil {
		return nil, fmt.Errorf("newsapi: http: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("newsapi: HTTP %d", resp.StatusCode)
	}

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("newsapi: read body: %w", err)
	}

	var apiResp newsAPIResponse
	if err := json.Unmarshal(body, &apiResp); err != nil {
		return nil, fmt.Errorf("newsapi: json parse: %w", err)
	}

	if apiResp.Status != "ok" {
		return nil, fmt.Errorf("newsapi: API returned status=%q", apiResp.Status)
	}

	fetchedAt := time.Now().UTC().Format(time.RFC3339)

	result := make([]RssItem, 0, len(apiResp.Articles))
	for _, a := range apiResp.Articles {
		title := a.Title
		if title == "" {
			title = a.Description
		}
		pubAt := ""
		if a.PublishedAt != "" {
			// NewsAPI returns RFC-3339 already; normalise for safety
			pubAt = normalizeRFC(a.PublishedAt)
		}

		ri := RssItem{
			ID:          dedupKey(a.URL, title, newsAPISource),
			Title:       title,
			URL:         a.URL,
			PublishedAt: pubAt,
			FetchedAt:   fetchedAt,
			Source:      newsAPISource,
		}
		result = append(result, ri)
	}
	return result, nil
}
